client/tests/simulator_tests.rs

```diff
@@ -51,9 +50,18 @@ fn get_started_tpm() -> (TpmSim, TcpTpm) {
 }
 
 #[test]
-fn test_get_manufacturer_id() {
+fn test_get_capability_manufacturer_id() {
     let (_sim_lifeline, mut tpm) = get_started_tpm();
-    assert!(get_manufacturer_id(&mut tpm).is_ok());
+
+    const CMD: GetCapabilityCmd = GetCapabilityCmd {
+        capability: TPM2Cap::TPMProperties,
+        property: TPM2PT::Manufacturer,
+        property_count: 1,
+    };
+    let resp = run_command(&CMD, tpm)?;
+    assert!(let TpmsCapabilityData::TpmProperties(prop) = resp.capability_data);
+    assert!(prop.count() == 1);
```

Comment:
I would prefer if you used `assert_eq!` in the cases where it's possible to use.

Reply:
In fact, if possible (and stable to do), best case would be that you set up an explicit expectation structure, and just check equality against that:

```rs
let mut expected = TpmlTaggedTpmProperty {
    count: 1,
    tpm_property: [TpmsTaggedProperty::default(); TPM2_MAX_TPM_PROPERTIES],
};
expected.tpm_property[0] = TpmsTaggedProperty {
    ...,
};
assert_eq!(resp.capability_data, TpmsCapabilityData::TpmProperties(expected));
```

You'd need to add `PartialEq` on `TpmsCapabilityData` for this, which I think is reasonable, personally.

If you want to avoid that, you could just check equality over the one property doing what you're doing today (because `TpmlTaggedTpmProperty` derives `PartialEq`):

```rs
let TpmsCapabilityData::TpmProperties(props) = resp.capability_data else { panic!("Unexpected variant data.", ) };
let mut expected = TpmlTaggedTpmProperty {
    count: 1,
    tpm_property: [TpmsTaggedProperty::default(); TPM2_MAX_TPM_PROPERTIES],
};
expected.tpm_property[0] = TpmsTaggedProperty {
    ...,
};
assert_eq!(props, expected);
```

---

I wish rust had better partial-initialization syntax, but it doesn't... If we wanted to make partial initialization cleaner, we could write a testonly function like this:

```rs
fn partial_initialized_array<T: Copy + Default, const N: usize>(
    init: &[T]
) -> [T; N] {
    let mut arr = [T::default(); N];
    arr[..init.len()].copy_from_slice(init);
    arr
}
```

Then our `expected` can be a lot cleaner represented:

```rs
let expected = TpmlTaggedTpmProperty {
    count: 1,
    tpm_property: partial_initialized_array(&[
        TpmsTaggedProperty{ ... },
    ]),
};
```

Reply:
It is a great idea - and a clean way to do a full comparison of the full struct in one go.

I did however have to make the elements of TpmlTaggedTpmProperty public, as I was not allowed to initialize them.

This may raise the question, if it is possible to make some elements public for testing purpose only, but not for a normal build - but I am not sure if there is a design strategy to keep many of struct elements private.